src/learner-record/index.ts

```diff
@@ -19,6 +27,64 @@ export class LearnerRecord {
 		this._inviteFactory = inviteFactory
 	}
 
+	async patchModuleRecord(jsonPatch: JsonPatch[], moduleRecordId: number) {
+		try {
+			this.logger.debug(`Patching module record with ID ${moduleRecordId}`)
+			const res: ModuleRecord = await this._restService.patchWithJsonPatch(`/module_records/${moduleRecordId}`, jsonPatch)
+			return plainToInstance(ModuleRecord, res)
+		} catch (e) {
+			throw new Error(`An error occurred when trying to patch the module record: ${e}`)
+		}
+	}
+	
+	async createModuleRecord(moduleRecord: ModuleRecordInput): Promise<ModuleRecord> {
+		try {
+			this.logger.debug(`Creating module record for module ID ${moduleRecord.moduleId} and user ID ${moduleRecord.userId}`)
+			const res: ModuleRecord = await this.restService.post('/module_records', moduleRecord)
+			return plainToInstance(ModuleRecord, res)
+		} catch (e) {
+			throw new Error(`An error occurred when trying to create the module record: ${e}`)
+		}
+	}
+
+	async patchCourseRecord(jsonPatch: JsonPatch[], userId: string, courseId: string) {
+		try {
+			this.logger.debug(`Patching course record for course ID ${courseId} and user ID ${userId}`)
+			const res: CourseRecord = await this._restService.patchWithJsonPatch(`/course_records?courseId=${courseId}&userId=${userId}`, jsonPatch)
+			return plainToInstance(CourseRecord, res)
+		} catch (e) {
+			throw new Error(`An error occurred when trying to patch the course record: ${e}`)
+		}
+	}
+	
+	async createCourseRecord(courseRecord: CourseRecordInput) {
+		try {
+			this.logger.debug(`Creating course record for course ID ${courseRecord.courseId}`)
+			const res: CourseRecord = await this.restService.post('/course_records', courseRecord)
+			return plainToInstance(CourseRecord, res)
+		} catch (e) {
+			throw new Error(`An error occurred when trying to create the course record: ${e}`)
+		}
+	}
+	
+	async getCourseRecord(courseId: string, userId: string): Promise<CourseRecord|undefined> {
+		try {
+			this.logger.debug(`Getting course record for course ID ${courseId} and user ID ${userId}`)
+			const data: CourseRecordResponse = await this._restService.get(`/course_records?courseId=${courseId}&userId=${userId}`)
+			const courseRecords = plainToInstance(CourseRecordResponse, data).courseRecords
+			let courseRecord
+			if (courseRecords.length === 1) {
+				courseRecord = plainToInstance(CourseRecord, courseRecords[0])
+			} else if (courseRecords.length > 1) {
+				this.logger.warn(`Course record for course ID ${courseId} and user ID ${userId} returned a result set greater than 1`)
+				courseRecord = plainToInstance(CourseRecord, courseRecords[0])
```

Comment:
A bit of repetition here. You could maybe return this outside the `if` statement, and keep the `warn` log entry only?

Reply:
Can't think of an elegant way to do this that preserves the `undefined` return 🤔

Reply:
If the second argument in `plainToInstance` is `undefined` (eg. getting the first instance of an empty array), then the result of `plainToInstance` will also be `undefined` (just tested it). So I was thinking something like this:

```ts
const courseRecords = plainToInstance(CourseRecordResponse, data).courseRecords

if (courseRecords.length > 1){
    this.logger.warn(`Course record for course ID ${courseId} and user ID ${userId} returned a result set greater than 1`)
}

// 👇 if `courseRecords` is an empty array (thus `courseRecords[0]` will be undefined) this will also return `undefined` 
let courseRecord = plainToInstance(CourseRecord, courseRecords[0]) 

return courseRecord
```

Reply:
Got ya! I had no idea you could do that, will get that in now